### PR TITLE
Miscellaneous format stuff

### PR DESCRIPTION
**Pull request by rmkaplan**
MACHINEINDEPENDENT using Interlisp readtable, WRITEFILE setting external format
\EXTERNALFORMAT can get its external format from a reader environment

### CODE REVIEW

**Review by masinter — CHANGES_REQUESTED** [review]
The function \THROUGHIN contains an extra paren around (BYTECOUNTVAR ...)

the "calls" ccode signature changed

File: library/UNICODE [hunk 1]
```diff
@@ -155,61 +161,57 @@
...
+
+    (* ;; "\BACKFILEPTR is NIL at beginning of FILE, do nothing.")
+
+    (* ;; "Common for big-ending and little-ending")
+
+    (IF (NOT (\BACKFILEPTR STREAM))
+      ELSEIF (\BACKFILEPTR STREAM)
+        THEN (AND BYTECOUNTVAR (SET BYTECOUNTVAR (IPLUS BYTECOUNTVAL 2)))
+      ELSE (AND BYTECOUNTVAR (SET BYTECOUNTVAR (ADD1 BYTECOUNTVAL])
 )
```
[the remark this answers is not in the document]
Comment (masinter) [reply library/UNICODE:175]: shouldn't \THROUGHBACKCCODE use SUB1 rather than ADD1 ?

**Comment by rmkaplan** [comment]
The inherited behavior is that these counters are counting down, which is not that intuitive.  It means that backing is positive.

Presumably this was done so that loops like in PFCOPYBYTES can test for 0 at the end.  

I’m going to make another little pass, after this merge, over these functions to build in the name of a SPECVAR (e.g. *BYTECOUNTVAR* that is bound in the \INCCODE and \BACKCCODE functions, so that all of these format implementations can do SETQ instead of SET to just that known variable.  They can also then do the intuitive thing of setting just the delta, the number of bytes that the file pointer has advanced (negative for backing).

The two generic funtions will update the caller’s actual counting variable, with a difference instead of a plus.

On the same pass, I’m going to move the character position stream setting up to the generic \OUTCHAR, so that it is not something that individual formats have to worry about—they just have to put out their bytes.  Less room for bugs because things are on the wrong side of the interface.

File: library/UNICODE (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmkaplan) [reply library/UNICODE:173]: Commented in email, SUB1 is correct (but not intuitive).

File: sources/MACHINEINDEPENDENT [hunk 2]
```diff
@@ -245,238 +246,1048 @@
...
+                                                                                  FILE T))
+       finally
+       (RETURN
+        (if (FMEMB FULL LOADEDFILELST)
+            then FULL
+          else (LET* [(ROOT (ROOTFILENAME FULL))
+                          (DATES (GETPROP ROOT 'FILEDATES))
+                          (FILEPROP (GETPROP ROOT 'FILE]
+                         (if [AND DATES (if (EQ (FILENAMEFIELD FULL 'EXTENSION)
+                                                        COMPILE.EXT)
```
Comment (masinter) [inline sources/MACHINEINDEPENDENT:266]: No DFASLs?

**Review by masinter — COMMENTED** [review]
Right now, you pass in two values: the name of the symbol and the old value (a number). If the variable name is set, then the value is non-NIL (a number).

 If you wanted to optimize the general case, instead of passing down the name of the \#chars variable, you could just always use, say, \\ncharsvar. Then you would still pass in the value  but use it also to test

(AND NCHARSVALUE (SETQ \\NCHARSVAR (IPLUS NCHARSVALUE 1))

maybe I'm not explaining this well.

**Comment by rmkaplan** [comment]
I think the description I put in (somewhere) is somewhat cleaner.

The name of the specvar is known, the only argument is a flag that says whether it needs to be set.  And the setting is just the number of bytes locally read or unread.  The generic function does the binding of the known specvar, and uses the increment-value to update the application-level variable.

At a higher level we could say that all applications that want to do keep track must bind and set the same known variable.  But I think I would do that still by having one known variable that is bound by the generic functions and set to the byte increment by the implementation functions, and then it would be a separate question of how the generic functions communicate with their environment.

But also see my issue about substreams and a possible SETEOFPTR, maybe this counting business is just at the wrong level, that the stream itself should provide an efficient way of tracking the character-to-byte corruption and obviate the need for the avoiding-large-numbers optimization of FILEPOS.

**Review by masinter — APPROVED** [review]

**Comment by masinter** [comment]
let me know when you're done with this pass or do you want to make another pass.
I'll then check out the conflicts.

**Pull request merged**